### PR TITLE
diff: refine code && add unit test for sharding tables

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 
1. refine code, rows should be closed after select, or it may cause blocked in next select.
2. add unit test for sharding tables

### What is changed and how it works?
1. scan rows after select data, and then close rows
2. diff more than one source table with one target table in unit test

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: pkg/dbutil/index.go [hunk 1]
```diff
@@ -42,24 +42,24 @@ func ShowIndex(ctx context.Context, db *sql.DB, schemaName string, table string)
 	defer rows.Close()
 
 	for rows.Next() {
-		fields, _, err1 := ScanRow(rows)
+		fields, err1 := ScanRow(rows)
 		if err1 != nil {
 			return nil, errors.Trace(err1)
 		}
-		seqInIndex, err1 := strconv.Atoi(string(fields["Seq_in_index"]))
+		seqInINdex, err1 := strconv.Atoi(string(fields["Seq_in_index"].Data))
```
Comment (kennytm) [inline pkg/dbutil/index.go]:
```suggestion
		seqInIndex, err1 := strconv.Atoi(string(fields["Seq_in_index"].Data))
```

File: pkg/diff/chunk.go [hunk 2]
```diff
@@ -201,10 +201,6 @@ func (s *randomSpliter) split(table *TableInstance, columns []*model.ColumnInfo,
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	if cnt == 0 {
-		log.Infof("no data found in %s.%s", table.Schema, table.Table)
-		return nil, nil
```
Comment (WangXiangUSTC) [inline pkg/diff/chunk.go:206]: if count is 0, also generate a chunk for this table

**Comment by WangXiangUSTC** [comment]
@GregoryIan @csuzhangxc  PTAL

File: pkg/diff/diff_test.go [hunk 3]
```diff
@@ -246,16 +270,46 @@ func generateData(dbConn *sql.DB, dbCfg dbutil.DBConfig, sourceTable, targetTabl
...
 	if err != nil {
 		return err
 	}
 
+	conditions := make([]string, 0, 3)
+	if randomValueNum == 0 {
+		conditions = append(conditions, "true")
+	} else {
+		conditions = append(conditions, fmt.Sprintf("e < %s", values[0]))
+		for i := range values {
```
Comment (csuzhangxc) [inline pkg/diff/diff_test.go]: `for i := 0; i < len(values)-1; i++ {`, then no `break` needed?

File: pkg/diff/diff_test.go [hunk 4]
```diff
@@ -246,16 +270,46 @@ func generateData(dbConn *sql.DB, dbCfg dbutil.DBConfig, sourceTable, targetTabl
 	importer.DoProcess(cfg)
 
 	// generate data for target table
```
Comment (csuzhangxc) [inline pkg/diff/diff_test.go]: update this comment.

**Comment by WangXiangUSTC** [comment]
@csuzhangxc PTAL again

**Comment by csuzhangxc** [comment]
LGTM

File: pkg/diff/chunk.go (hunk 2 above)
Comment (IANTHEREAL) [reply pkg/diff/chunk.go:206]: why？

File: pkg/diff/diff.go [hunk 5]
```diff
@@ -298,13 +299,11 @@ func (t *TableDiff) checkChunkDataEqual(ctx context.Context, checkJobs []*CheckJ
 	return equal, nil
 }
 
-func (t *TableDiff) compareRows(sourceRows map[string]*sql.Rows, targetRows *sql.Rows, orderKeyCols []*model.ColumnInfo) (bool, error) {
+func (t *TableDiff) compareRows(sourceRows map[string][]map[string]*dbutil.ColumnData, targetRows []map[string]*dbutil.ColumnData, orderKeyCols []*model.ColumnInfo) (bool, error) {
```
Comment (IANTHEREAL) [inline pkg/diff/diff.go:302]: why change it？

File: pkg/diff/chunk.go (hunk 2 above)
Comment (WangXiangUSTC) [reply pkg/diff/chunk.go:206]: empty in target, but may be have data in source

File: pkg/diff/diff.go (hunk 5 above)
Comment (WangXiangUSTC) [reply pkg/diff/diff.go:302]: we should read data from sql.Rows and then close the rows.

Comment (IANTHEREAL) [reply pkg/diff/diff.go:302]: if use `rows`, we can read it while we need, now we should read all once, right?

but you can keep it, while you need to refine mem usage, you should change it back

**Comment by IANTHEREAL** [comment]
LGTM

File: pkg/diff/diff.go (hunk 5 above)
Comment (WangXiangUSTC) [reply pkg/diff/diff.go:302]: ok

**Pull request merged**